File: 1143-longest-common-subsequence/1143-longest-common-subsequence.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        return self.maxCommonLength(text1, text2, 0, 0, dict())
        
    def maxCommonLength(self, text1, text2, start1, start2, memo):
        if start1 == len(text1) or start2 == len(text2):
            return 0
        key = tuple([start1, start2])
        if key in memo:
            return memo[key]
        res = max(self.maxCommonLength(text1, text2, start1, start2 + 1, memo), self.maxCommonLength(text1, text2, start1 + 1, start2, memo))
        if text1[start1] == text2[start2]:
            res = max(res, self.maxCommonLength(text1, text2, start1 + 1, start2 + 1, memo) + 1)
        memo[key] = res
        return res
            
        

File: 1143-longest-common-subsequence/1143-longest-common-subsequence.java
class Solution {
    public int longestCommonSubsequence(String text1, String text2) {
        int text1Length = text1.length();
        int text2Length = text2.length();
        
        int[][] longestCommon = new int[text1Length + 1][text2Length + 1];
        
        int ans = 0;
        
        for (int p1 = 1; p1 < text1Length + 1; p1++)
            for (int p2 = 1; p2 < text2Length + 1; p2++){
                char text1Char = text1.charAt(p1 - 1);
                char text2Char = text2.charAt(p2 - 1);
                
                if (text1Char == text2Char){
                    longestCommon[p1][p2] = Math.max(longestCommon[p1][p2], longestCommon[p1 - 1][p2 - 1] + 1);
                }else{
                    longestCommon[p1][p2] = Math.max(longestCommon[p1][p2], longestCommon[p1 - 1][p2]);
                    longestCommon[p1][p2] = Math.max(longestCommon[p1][p2], longestCommon[p1][p2 - 1]);
                }
                
                ans = Math.max(longestCommon[p1][p2], ans);
            }
        
        return ans;
    
    }
}